File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/functional_type/functional_type.swift
import FunctionalType
import Testing
import ref_types
import optional_types
import primitive_types
import collection_types
import data
import receivers

@Test
func testCallingClosureSentToKotlin() throws {
    var i: Int = 0
    foo_consume_simple {
      i += 1
    }
    try #require(i == 0)
    call_consumed_simple_block()
    try #require(i == 1)
    call_consumed_simple_block()
    try #require(i == 2)
}

@Test
func testSimpleProduceBlock() throws {
    let block = foo_produce_simple()
    try #require(foo_produce_simple_counter == 0)
    block()
    try #require(foo_produce_simple_counter == 1)
    block()
    try #require(foo_produce_simple_counter == 2)
}

@Test
func testSimpleProduceBlockReturningInt() throws {
    let block = foo_produce_simple_int()
    try #require(block() == 1)
    try #require(block() == 2)
    try #require(block() == 3)
}

@Test
func testFunctionalTypeTypealias() throws {
    var i: Int = 0
    save_typealiased_callback {
      i += 2
    }
    try #require(i == 0)
    call_saved_callback()
    try #require(i == 2)
}

@Test
func testBlockWithRefType() throws {
    for _ in 0...1_000_000 {
        var lastB: Bar? = nil
        var receivedB: Bar? = nil
        saveRefBlock { b in
            lastB = b
            return Bar(i: b.i+1)
        }
        receivedB = callRefBlock(with: Bar(i: 0))
        try #require(receivedB!.i == 1)
        try #require(lastB!.i == 0)
    }
}

@Test
func testBlockWithOptRefType() throws {
    var lastB: Bar? = nil
    var receivedB: Bar? = nil
    saveOptRefBlock { b in
        lastB = b
        return b.flatMap { it in Bar(i: it.i+1) }
    }
    receivedB = callOptRefBlock(with: Bar(i: 0))
    try #require(receivedB!.i == 1)
    try #require(lastB!.i == 0)

    receivedB = callOptRefBlock(with: nil)
    try #require(receivedB == nil)
    try #require(lastB == nil)
}

@Test
func testProduceBlockWithOptRefType() throws {
    let block = produceOptionalId()
    var receivedB: Bar? = nil

    receivedB = block(Bar(i: 0))
    try #require(receivedB!.i == 1)
    try #require(last_seen_bar_by_produceOptionalId!.i == 0)

    receivedB = block(nil)
    try #require(receivedB == nil)
    try #require(last_seen_bar_by_produceOptionalId == nil)
}

@Test
func testBlockWithOptPrimType() throws {
    var last: Int32? = nil
    var received: Int32? = nil
    saveOptPrimBlock { i in
        last = i
        return i?.advanced(by: 1)
    }
    received = callOptPrimBlock(with: 0)
    try #require(received == 1)
    try #require(last == 0)

    received = callOptPrimBlock(with: nil)
    try #require(received == nil)
    try #require(last == nil)
}

@Test
func testBlockWithPrimType() throws {
    var last: Int8? = nil
    var received: Int8? = nil
    savePrimBlock { byte in
        last = byte
        return byte+1
    }

    received = callPrimBlock(with: 0)
    try #require(received == 1)
    try #require(last == 0)
}

@Test
func testBlockWithListType() throws {
    var last: [Int32] = []
    var received: [Int32] = []
    saveListBlock { it in
        last = it
        return it.reversed()
    }

    received = callListBlock(with: [1, 2, 3])
    try #require(received == [3, 2, 1])
    try #require(last == [1, 2, 3])
}

@Test
func testProduceBlockWithListType() throws {
    let block = produceListId()
    var received: [Int32] = []

    received = block([1, 2, 3])
    try #require(received == [3, 2, 1])
    try #require(last_seen_bar_by_produceListId == [1, 2, 3])
}

@Test
func testFunctionWithIntReceiver() throws {
    var received: Int32? = nil
    fooReceiverInt { it in
        received = it
    }
    try #require(received == 5)
}

@Test
func testFunctionWithStringReceiver() throws {
    var received: String? = nil
    fooReceiverString { it in
        received = it
    }
    try #require(received == "hello")
}

@Test
func testFunctionWithBarReceiver() throws {
    var received: Bar? = nil
    fooReceiverBar { it in
        received = it
    }
    try #require(received!.i == 5)
}

@Test
func testFunctionWithListReceiver() throws {
    var received: [Int32]? = nil
    fooReceiverList { it in
        received = it
    }
    try #require(received!.count == 3)
    try #require(received == [1, 2, 3])
}

@Test
func testProduceBlockWithStringReceiver() throws {
    let block = produceWithStringReceiver()
    var received: String = ""

    received = block("hello")
    try #require(received == "hellohello")
    try #require(last_seen_bar_by_produceWithStringReceiver == "hello")
}

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/functional_type/functional_type.kt
// KIND: STANDALONE
// MODULE: FunctionalType
// FILE: functional_type_simple_consume.kt

private lateinit var block: ()->Unit
fun call_consumed_simple_block() = block()
fun foo_consume_simple(block_input: ()->Unit): Unit {
    block = block_input
}

// FILE: functional_type_simple_produce.kt

var foo_produce_simple_counter: Int = 0
fun foo_produce_simple(): ()->Unit = {
    foo_produce_simple_counter += 1
}

var foo_produce_simple_int_counter: Int = 0
fun foo_produce_simple_int(): ()->Int = {
    foo_produce_simple_int_counter += 1
    foo_produce_simple_int_counter
}

// FILE: functional_type_typealias.kt

typealias callback = () -> Unit
private var callback_property: callback = {}
fun save_typealiased_callback(input: callback) {
    callback_property = input
}
fun call_saved_callback() = callback_property()

// MODULE: ref_types(data)
// EXPORT_TO_SWIFT
// FILE: ref_types.kt

private lateinit var ref_block: (Bar) -> Bar

fun saveRefBlock(block: (Bar) -> Bar) {
    ref_block = block
}
fun callRefBlock(with: Bar): Bar = ref_block(with)

// MODULE: optional_types(data)
// EXPORT_TO_SWIFT
// FILE: optional_types.kt

private lateinit var optional_ref_block: (Bar?) -> Bar?

fun saveOptRefBlock(block: (Bar?) -> Bar?) {
    optional_ref_block = block
}
fun callOptRefBlock(with: Bar?): Bar? = optional_ref_block(with)

private lateinit var optional_prim_block: (Int?) -> Int?

fun saveOptPrimBlock(block: (Int?) -> Int?) {
    optional_prim_block = block
}
fun callOptPrimBlock(with: Int?): Int? = optional_prim_block(with)

// FILE: functional_type_produce_optional_id.kt

var last_seen_bar_by_produceOptionalId: Bar? = null
fun produceOptionalId(): (Bar?)->Bar? = {
    last_seen_bar_by_produceOptionalId = it
    if (it != null) Bar(it.i + 1) else null
}

// MODULE: primitive_types
// EXPORT_TO_SWIFT
// FILE: primitive_types.kt

private lateinit var prim_block: (Byte) -> Byte

fun savePrimBlock(block: (Byte) -> Byte) {
    prim_block = block
}
fun callPrimBlock(with: Byte): Byte = prim_block(with)

// MODULE: collection_types
// EXPORT_TO_SWIFT
// FILE: collection_types.kt

private lateinit var list_block: (List<Int>) -> List<Int>

fun saveListBlock(block: (List<Int>) -> List<Int>) {
    list_block = block
}
fun callListBlock(with: List<Int>): List<Int> = list_block(with)

// FILE: functional_type_produce_collection.kt

var last_seen_bar_by_produceListId: List<Int>? = null
fun produceListId(): (List<Int>)->List<Int> = {
    last_seen_bar_by_produceListId = it
    it.asReversed()
}

// MODULE: receivers(data)
// EXPORT_TO_SWIFT
// FILE: functional_type_with_receiver.kt

fun fooReceiverInt(i: Int.()->Unit): Unit = with(5) { i() }

fun fooReceiverString(i: String?.()->Unit): Unit = with("hello") { i() }

fun fooReceiverBar(i: Bar.()->Unit): Unit = with(Bar(5)) { i() }

fun fooReceiverList(i: List<Int>.()->Unit): Unit = with(listOf(1, 2, 3)) { i() }

// FILE: functional_type_produce_withStringReceiver.kt

var last_seen_bar_by_produceWithStringReceiver: String? = null
fun produceWithStringReceiver(): String.()->String = {
    last_seen_bar_by_produceWithStringReceiver = this
    "$this$this"
}

// MODULE: data
// EXPORT_TO_SWIFT
// FILE: data.kt

class Bar(var i: Int)
